## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

pd.options.display.float_format = '{:.2f}'.format

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()


client_id   order_id  order_week  order_year  unit_price  unit_cost  \
count   54639.00   54639.00    54639.00    54639.00    54639.00   54639.00   
mean    54837.87 5470190.03       11.36     2022.99      136.27      99.45   
std     25487.44 2599806.93        7.02        0.08      183.87     133.16   
min     10033.00 1000886.00        1.00     2022.00        0.01       0.01   
25%     33593.00 3196372.00        6.00     2023.00       20.80      14.84   
50%     53305.00 5496966.00       11.00     2023.00       68.31      49.89   
75%     78498.00 7733869.00       17.00     2023.00      173.16     125.57   
max     99984.00 9998480.00       52.00     2023.00     1396.23     846.27   

       unit_weight        qty  line_number  
count     54639.00   54639.00     54639.00  
mean          5.00     570.26         2.98  
std           5.33   18795.52         2.44  
min           0.00       0.00         0.00  
25%           1.45      32.00         1.00  
50%           3.24      68.00         3.00  
75%           6.89     170.00         5.00  
max          46.43 3958244.00         9.00

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?

#Filter to "consumables"
consumables_df = df.loc[df['category'] == 'consumables']

#Find the most entries within that category
consumables_df['subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
top_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top_clients_list = top_clients.index.tolist()

top_clients_list

[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the client with the most entries order?
most_total_units = df[df['client_id'] == 33615]['qty'].sum()
most_total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  subtotal  
0     1096.80     762.71         7.50  105            1 115164.00  
1       24.95      15.09         1.49   21            0    523.95  
2       13.52       7.86         1.68   39            6    527.28  
3       36.42      24.85         1.23   29            3   1056.18  
4      195.10     108.17        46.43   20            1   3902.00

In [11]:
# Create a column for shipping price. Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

#Create a column for the Total Weight of the Order
df['Total Weight'] = df['unit_weight'] * df['qty']

# Calculate shipping price based on total weight
df['Shipping Price'] = df['Total Weight'] * 7  
df.loc[df['Total Weight'] <= 50, 'Shipping Price'] = df['Total Weight'] * 10 

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number  subtotal  \
0        1096.80     762.71         7.50   105            1 115164.00   
1          24.95      15.09         1.49    21            0    523.95   
2          13.52       7.86         1.68    39            6    527.28   
3          36.42      24.85         1.23    29            3   1056.18   
4         195.10     108.17        46.43    20            1   3902.00   
...          ...        ...          ...   ...          ...       ...   
54634      83.13      51.60         2.25    33            8   2743.29   
54635     206.59     175.46        11.70    47            0   9709.73   
54636      65.66      57.31         4.16   475            6  31188.50   
54637       1.48       1.22        18.04   112            0    165.76   
54638       3.01       1.61         2.07  1031            5   3103.31   

       Total Weight  Shipping Price  
0            787.50         5512.50  
1             31.29          312.90  
2             65.52          458.64  
3             35.67          356.70  
4            928.60         6500.20  
...             ...             ...  
54634         74.25          519.75  
54635        549.90         3849.30  
54636       1976.00        13832.00  
54637       2020.48        14143.36  
5

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 0.0925

calc_sales_tax = (df['subtotal'] + df['Shipping Price']) * sales_tax_rate

df['Total Price'] = df['subtotal'] + df['Shipping Price'] + calc_sales_tax

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number  subtotal  Total Weight  Shipping Price  \
0             7.50   105            1 115164.00        787.50         5512.50   
1             1.49    21            0    523.95         31.29          312.90   
2             1.68    39            6    527.28         65.52          458.64   
3             1.23    29            3   1056.18         35.67          356.70   
4            46.43    20            1   3902.00        928.60         6500.20   
...            ...   ...          ...       ...           ...             ...   
54634         2.25    33            8   2743.29         74.25          519.75   
54635        11.70    47            0   9709.73        549.90         3849.30   
54636         4.16   475            6  31188.50       1976.00        13832.00   
54637        18.04   112            0    165.76       2020.48        14143.36   
54638         2.07  1031            5   3103.31       2134.17        14939.19   

       Total Price  
0        131839.08  
1           914.26  
2          1077.12  
3          1543.57  
4         11364.40  
...            ...  
54634      3564.87  
54635     14813.24  
54636     49184.90  
54637     15632.7

In [13]:
# Create a column for the cost of each line using unit cost, qty, and shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['Total Cost'] = df['unit_cost'] * df['qty'] + df['Shipping Price']

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  subtotal  Total Weight  Shipping Price  Total Price  \
0                1 115164.00        787.50         5512.50    131839.08   
1                0    523.95         31.29          312.90       914.26   
2                6    527.28         65.52          458.64      1077.12   
3                3   1056.18         35.67          356.70      1543.57   
4                1   3902.00        928.60         6500.20     11364.40   
...            ...       ...           ...             ...          ...   
54634            8   2743.29         74.25          519.75      3564.87   
54635            0   9709.73        549.90         3849.30     14813.24   
54636            6  31188.50       1976.00        13832.00     49184.90   
54637            0    165.76       2020.48        14143.36     15632.71   
54638            5   3103.31       2134.17        14939.19     19711.43   

       Total Cost  
0        85597.05  
1          629.79  
2          765.18  
3         1077.35  
4         8663.60  
...           ...  
54634     2222.55  
54635    12095.92  
54636    41054.25  
54637    14280.00  
54638    16599.10  

[54639 rows x 23 columns]

In [14]:
# Create a column for the profit of each line using line cost and line price
df['Profit'] = df['Total Price'] - df['Total Cost']

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

       subtotal  Total Weight  Shipping Price  Total Price  Total Cost  \
0     115164.00        787.50         5512.50    131839.08    85597.05   
1        523.95         31.29          312.90       914.26      629.79   
2        527.28         65.52          458.64      1077.12      765.18   
3       1056.18         35.67          356.70      1543.57     1077.35   
4       3902.00        928.60         6500.20     11364.40     8663.60   
...         ...           ...             ...          ...         ...   
54634   2743.29         74.25          519.75      3564.87     2222.55   
54635   9709.73        549.90         3849.30     14813.24    12095.92   
54636  31188.50       1976.00        13832.00     49184.90    41054.25   
54637    165.76       2020.48        14143.36     15632.71    14280.00   
54638   3103.31       2134.17        14939.19     19711.43    16599.10   

        Profit  
0     46242.03  
1       284.47  
2       311.94  
3       466.22  
4      2700.80  
...        ...  
54634  1342.32  
54635  2717.32  
54636  8130.65  
54637  1352.71  
54638  3112.33  

[54639 rows x 24 columns]

In [15]:
# Clean data formats
df['Shipping Price'] = df['Shipping Price'].astype(float)
df['Total Price'] = df['Total Price'].astype(float)
df['Total Cost'] = df['Total Cost'].astype(float)
df['Profit'] = df['Profit'].astype(float)

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

       subtotal  Total Weight  Shipping Price  Total Price  Total Cost  \
0     115164.00        787.50         5512.50    131839.08    85597.05   
1        523.95         31.29          312.90       914.26      629.79   
2        527.28         65.52          458.64      1077.12      765.18   
3       1056.18         35.67          356.70      1543.57     1077.35   
4       3902.00        928.60         6500.20     11364.40     8663.60   
...         ...           ...             ...          ...         ...   
54634   2743.29         74.25          519.75      3564.87     2222.55   
54635   9709.73        549.90         3849.30     14813.24    12095.92   
54636  31188.50       1976.00        13832.00     49184.90    41054.25   
54637    165.76       2020.48        14143.36     15632.71    14280.00   
54638   3103.31       2134.17        14939.19     19711.43    16599.10   

        Profit  
0     46242.03  
1       284.47  
2       311.94  
3       466.22  
4      2700.80  
...        ...  
54634  1342.32  
54635  2717.32  
54636  8130.65  
54637  1352.71  
54638  3112.33  

[54639 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
#total_client_spend = df.loc[df['order_id'] == 2742071, 'Total Price'].sum()
#total_client_spend



152811.897075

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
sorted_qty_df = df.sort_values(by='qty', ascending=False)
highest_qty_client_list = sorted_qty_df['client_id'].head(5).tolist()

client_spend = df[df['client_id'].apply(lambda x: x in highest_qty_client_list)].groupby('client_id')['Total Price'].sum()
client_spend_currency = client_spend.map('${:,.2f}'.format)
client_spend_currency

client_id
30143     $33,328,849.83
77354     $71,942,651.78
78965    $655,923,670.06
85931    $237,417,001.42
86721    $181,266,297.07
Name: Total Price, dtype: object

In [18]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. Sort by total profit.

# Filter the DataFrame for the top 5 clients
top_5_client_df = df[df['client_id'].apply(lambda x: x in highest_qty_client_list)]

# Calculate total units purchased, total shipping price, total revenue, and total profit for each client
client_totals = top_5_client_df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('Shipping Price', 'sum'),
    total_revenue=('Total Price', 'sum'),
    total_profit=('Profit', 'sum')
)

# Sort the summary DataFrame by total profit
summary_df = client_totals.sort_values(by='total_profit', ascending=False)

summary_df


total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
78965                    4033302            5340756.88   655923670.06   
85931                     644423           10122549.19   237417001.42   
86721                    1010426           64335727.72   181266297.07   
77354                     506382           23814685.86    71942651.78   
30143                     846069            8981905.24    33328849.83   

           total_profit  
client_id                
78965      247352842.47  
85931      109294128.82  
86721       42974695.49  
77354       12432260.11  
30143       11250385.89

In [19]:
# Format the data and rename the columns to names suitable for presentation. Currency should be in millions of dollars.

summary_df = summary_df.rename(columns={
    'total_units_purchased': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Price',
    'total_revenue': 'Total Revenue',
    'total_profit': 'Total Profit'
})

summary_df=summary_df.applymap("${:,.2f}M".format)

summary_df


/var/folders/m6/qbpjxbts54s21jwmn73ttxyr0000gp/T/ipykernel_30534/2366734512.py:10: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  summary_df=summary_df.applymap("${:,.2f}M".format)


Total Units Purchased Total Shipping Price     Total Revenue  \
client_id                                                                
78965            $4,033,302.00M       $5,340,756.88M  $655,923,670.06M   
85931              $644,423.00M      $10,122,549.19M  $237,417,001.42M   
86721            $1,010,426.00M      $64,335,727.72M  $181,266,297.07M   
77354              $506,382.00M      $23,814,685.86M   $71,942,651.78M   
30143              $846,069.00M       $8,981,905.24M   $33,328,849.83M   

               Total Profit  
client_id                    
78965      $247,352,842.47M  
85931      $109,294,128.82M  
86721       $42,974,695.49M  
77354       $12,432,260.11M  
30143       $11,250,385.89M

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
# The df is already formatted this way.